README.md

```diff
@@ -4,7 +4,71 @@
 
 ## What is Tonatona?
 
-`Tonatona` is **meta** web application framework, which handles lots of annoying and boring tasks on real world web development such as environment variables, logging, etc...
+Tonatona is **meta** web application framework. It handles lots of boring tasks
+that are needed for real-world web development such as reading in values
+defined in environment variables, setting up logging, sending emails, accessing
+databases, etc.
 
-`Tonatona` **does not** provide explicit ways to realize core functionalities of web applications.
-It means that you can use your favorite framework based on `Network.Wai`! (e.g., Servant, Spock, Scotty, etc...)
+Tonatona **does not** provide the core functionalities of web applications,
+such as routing, request parsing, response building, etc.  Instead, you can
+use Tonatona with your favorite web framework.  Tonatona should work with Servant,
+Spock, Yesod, and others.
+
+Tonatona provides a plugin architecture so that anyone can add plugins
+implementing arbitrary functionality.  This repository contains many standard
+plugins that are helpful when writing Haskell applications.
+
+## How to use Tonatona
+
+(TODO)
+
+## Available Plugins
+
+(TODO)
+
+## Goals for Tonatona
+
+Tonatona is meant to make it easy to write web applications.  Its goal is to
+free the end-user from having to write anything but their application's
+business logic.
+
+We have the following goals and objectives in mind when developing Tonatona and
+the plugins.
+
+-   Tonatona should make it easy to get up and running with writing web
+    applications.  It is perfect when you are on a small team.
+
+    When you team or project grows larger, it may be worthwhile to move away
+    from Tonatona onto hand-written logging, DB access, etc.
+
+-   Make the end-user write a little boilerplate code up front in order to provide
+    ease-of-use when writing their own business logic.
+
+-   At runtime, plugins should be able to be configured by providing
+    environment variables.  Changing the value of the environment variables
+    should change the runtime behavior of your code.
+
+-   It should be possible to switch plugins just by rewriting the import statements.
+
+    For instance, it should be possible to go from using `tonatona-db-sqlite` to
+    `tonatona-db-postgresql` just by changing the import statement
+
+    ```haskell
+    import qualified Tonatona.Db.Sqlite as TonaDb
+    ```
+
+    to
+
+    ```haskell
+    import qualified Tonatona.Db.Postgresql as TonaDb
+    ```
+
+## Contributing
+
+Information about contributing new plugins can be found in
+[CONTRIBUTING.md](./CONTRIBUTING.md).
+
```

Comment:
Some people would want to develop new plugins on their own repository and publish to their own package.
Do you have any good idea when people should add new plugins to this repository or to their own repositories?

Reply:
Hmm, I'm not sure about this.

In general, I think that if a plugin is useful in general, it could be added to this repository.

Two examples of "plugin-based" projects written in Haskell are Servant and XMonad.  I'll describe a little how each works:

- XMonad

  XMonad is split into three packages/repositories, [`xmonad`](https://github.com/xmonad/xmonad), [`xmonad-contrib`](https://github.com/xmonad/xmonad-contrib), and [`xmonad-extras`](https://github.com/xmonad/xmonad-extras).  `xmonad` contains basic functionality for the window manager.  `xmonad-contrib` contains additional functionality that doesn't depend on external packages/libraries.  This has many modules implementing lots of functionality.  `xmonad-extras` contains additional functionality that depends on external packages and system libraries.

  `xmonad` is kept pretty small, while `xmonad-contrib`and `xmonad-extras` seem to accept any PR.  All three are kept under the XMonad Github Organization.  I think XMonad maintainers help maintain all three.

- Servant

  The main servant repository contains the following packages: `servant`, `servant-server`, `servant-docs`, `servant-client` (as well as a few others).

  Additional servant packages (like `servant-swagger` or `servant-checked-exceptions`) are all maintained by other people outside of the main servant repository.

  

I think we have a question as to whether or not we want to help maintain third-party plugins.  If so, then we should accept them in this repository.  If not, then we should make people maintain them separately.

Reply:
Thanks for great overview.
I agree with you.

Reply:
In commit 52bbfd2b1d I added an explanation as to which plugins will be accepted.  Does this sound good?

```
In general, new plugins will be accepted to Tonatona if they are widely useful.
For instance, a plugin adding support for a widely used database library will
probably be accepted, while a plugin adding support for a proprietary library
not widely used will probably not be accepted.

If your plugin is not accepted into this repository, you are free to support it
as a third-party repository, release it on Hackage, etc.  If you are using
Tonatona in a larger project, you will probably end up creating a few of your
own plugins!
```

Reply:
That sounds good!